cni/network/invoker_cns.go

```diff
@@ -130,97 +131,143 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
+						},
 					},
-				},
-				Routes: []*cniTypes.Route{
-					{
-						Dst: network.Ipv4DefaultRouteDstPrefix,
-						GW:  ncgw,
+					Interfaces: []*cniTypesCurr.Interface{
+						{
+							Mac: info.macAddress,
+						},
 					},
 				},
+				addressType:        cns.Secondary,
+				macAddress:         macAddress,
+				isDefaultInterface: info.isDefaultInterface,
+			}
+
+			if err = getRoutes(result.ipResult, info.routes); err != nil {
+				return IPAMAddResult{}, err
+			}
+
+			addResult.cniResults = append(addResult.cniResults, result)
+		default:
+			// set the NC Primary IP in options
+			// SNATIPKey is not set for ipv6
+			if net.ParseIP(info.ncPrimaryIP).To4() != nil {
+				addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
 			}
-		} else if net.ParseIP(info.podIPAddress).To16() != nil {
-			addResult.ipv6Result = &cniTypesCurr.Result{
-				IPs: []*cniTypesCurr.IPConfig{
-					{
+
+			ip, ncIPNet, err := net.ParseCIDR(info.podIPAddress + "/" + fmt.Sprint(info.ncSubnetPrefix))
+			if ip == nil {
+				return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
+			}
+
+			ncgw := net.ParseIP(info.ncGatewayIPAddress)
+			if ncgw == nil {
+				// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
+				if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualStackOverlay) && (invoker.ipamMode != util.Overlay) {
+					return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+				}
+
+				if net.ParseIP(info.podIPAddress).To4() != nil { //nolint:gocritic
+					ncgw, err = getOverlayGateway(ncIPNet)
+					if err != nil {
+						return IPAMAddResult{}, err
+					}
+				} else if net.ParseIP(info.podIPAddress).To16() != nil {
+					ncgw = net.ParseIP(overlayGatewayV6IP)
+				} else {
+					return IPAMAddResult{}, errors.Wrap(err, "No podIPAddress is found: %w")
+				}
+			}
+
+			// construct ipnet for result
+			resultIPnet := net.IPNet{
+				IP:   ip,
+				Mask: ncIPNet.Mask,
+			}
+
+			if ip := net.ParseIP(info.podIPAddress); ip != nil {
+				defaultCniResult := addResult.defaultCniResult.ipResult
+				if defaultCniResult == nil {
+					defaultCniResult = &cniTypesCurr.Result{}
+				}
+
+				defaultRouteDstPrefix := network.Ipv4DefaultRouteDstPrefix
+				if ip.To4() == nil {
+					defaultRouteDstPrefix = network.Ipv6DefaultRouteDstPrefix
+					addResult.ipv6Enabled = true
+				}
+
+				defaultCniResult.IPs = append(defaultCniResult.IPs,
+					&cniTypesCurr.IPConfig{
 						Address: resultIPnet,
 						Gateway: ncgw,
-					},
-				},
-				Routes: []*cniTypes.Route{
-					{
-						Dst: network.Ipv6DefaultRouteDstPrefix,
+					})
+				defaultCniResult.Routes = append(defaultCniResult.Routes,
+					&cniTypes.Route{
+						Dst: defaultRouteDstPrefix,
 						GW:  ncgw,
-					},
-				},
+					})
+
+				if err = getRoutes(defaultCniResult, info.routes); err != nil {
```

Comment:
can `getRoutes` return routes and then assign it to cniresult? updating inside the function doesn't look correct

Reply:
yup, updated